server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -52,25 +52,295 @@
+                    if(mid < numClusters - 1) {
+                        double valAfterMid = clusterCentroids.get(mid + 1);
+                        if (val < valAfterMid) {
+                            // val is between indices mid, (mid + 1)
+                            if (valAfterMid - val < val - midVal) {
+                                return mid + 1;
+                            } else {
+                                return mid;
+                            }
+                        }
+                    } else if(mid == numClusters - 1){
+                        // There are are no more clusters above mid so mid is the closest
+                        return mid;
+                    }
+
+                    left = mid + 1;
+                } else{ // midVal > val
+                    // Check if mid is the closest element. This is still possible because val is a double
+                    if(mid > 0) {
+                        double valBeforeMid = clusterCentroids.get(mid - 1);
+                        if (val > valBeforeMid) {
+                            // val is between indices (mid - 1), mid
+                            if (midVal - val < val - valBeforeMid) {
+                                return mid;
+                            } else {
+                                return mid - 1;
+                            }
+                        }
+                    } else if(mid == 0){
+                        // There are no more clusters below mid, so mid is the closest
+                        return mid;
+                    }
+
+                    right = mid - 1;
+                }
+            }
+
+            return left;
+        }
+
+        public void doClose(){
+            Releasables.close(clusterMaxes, clusterMins, clusterCentroids, clusterSizes);
+        }
+    }
+
     private final ValuesSource.Numeric valuesSource;
     private final DocValueFormat formatter;
 
-    private final int numBuckets; // Final number of buckets that should be produced by the coordinator node after reduction
-    private final int shardSize; // Max number of buckets that this aggregator should return for reduction
-    private final int cacheLimit; // Number of documents that are stored in memory before initial bucketing occurs
-    private int numCachedDocs;
-    private int avgBucketDistance;
+    // Aggregation parameters
+    private final int numBuckets;
+    private final int shardSize;
+    private final int cacheLimit;
 
-    /**
-     * "Cluster" refers to intermediate buckets during collection
-     */
     final BigArrays bigArrays;
-    private DoubleArray cachedValues;
-    private DoubleArray clusterMaxes;
-    private DoubleArray clusterMins;
-    private DoubleArray clusterCentroids;
-    private DoubleArray clusterSizes; // clusterSizes.get(i) will not be equal to bucketDocCount(i) when clusters are being merged
-    private int numClusters;
+    CacheValues cacheValues;
```

Comment:
Could this be `private CollectMode collect`? And you start with `CacheValues` and `close` that when it is done? Maybe `collectValue` could return `CollectMode` and `CacheValues` could return itself until it is time to transition and then it'd build its replacement and close itself? I'm *guessing* that'd be a bit easier to understand for me, mostly because I really can't keep a bunch of things in my head at once and sort of need to aggressively make abstractions.